algorithms/unit_tests/TestRandom.hpp

```diff
@@ -539,6 +539,57 @@ void test_duplicate_stream() {
   }
 }
 
+template <class ExecutionSpace, class GeneratorPool>
+struct compare_random_streams {
+  using ViewType = Kokkos::View<uint64_t**, ExecutionSpace>;
+
+  ViewType vals;
+  GeneratorPool rand_pool;
+  int samples;
+
+  compare_random_streams(ViewType vals_, GeneratorPool rand_pool_, int samples_)
+      : vals(vals_), rand_pool(rand_pool_), samples(samples_) {}
+
+  KOKKOS_INLINE_FUNCTION
+  void operator()(int i, std::size_t& mismatches) const {
+    typename GeneratorPool::generator_type rand_gen = rand_pool.get_state();
+
+    for (int k = 0; k < samples; k++)
+      if (vals(i, k) != rand_gen.urand64()) mismatches++;
+
+    rand_pool.free_state(rand_gen);
+  }
+};
+
+template <class ExecutionSpace, class Pool>
+void test_async_initialization() {
+  using ViewType = Kokkos::View<uint64_t**, ExecutionSpace>;
+
+  int n_streams = 1;
+  int samples   = 123456;
+
+  // use default execution space instance to generate reference values
+  Pool rand_pool_A(42);
+  ViewType vals_d("Vals", n_streams, samples);
+  Kokkos::parallel_for(Kokkos::RangePolicy<ExecutionSpace>(0, n_streams),
+                       generate_random_stream<ExecutionSpace, Pool>(
+                           vals_d, rand_pool_A, samples));
+
+  // generate same streams using explicit ExecutionSpace instance, and
+  // compare to reference values
+  ExecutionSpace exec{};
```

Comment:
Maybe use `partition_space` to create a different execution space instance than used above.
Either way, I think it's important to be explicit about the execution space instance for the `parallel_for` above. Whether you need the `fence` or not, depends on choosing a different execution space instance or not.

Reply:
done